File: solutions_1-10/Euler_3.py
# what is the largest prime factor of 600851475143 ? 

import math
import timeit


start = timeit.default_timer()

def is_prime(x):
    if x == 2:
        return True
    if x == 1 or x % 2 == 0:
        return False
    for i in range(3,int(math.sqrt(x)) + 1, 2):
        if x % i == 0:
            return False
    return True

def euler_3():
    prims = [i for i in xrange(1, 1000000) if is_prime(i) and 600851475143 % i == 0]
    return max(prims)

print "Answer: %s" % euler_3()
stop = timeit.default_timer()
print "Time: %f" % (stop - start)

File: Java/java_solutions_1-10/Euler_3.java
/*
/   Project Euler problem #3 Largest prime factor
/
/   Find the largest prime factor of the number 600851475143
/
*/

import java.util.stream.IntStream;
import static java.lang.Math.sqrt;
import static java.lang.Math.ceil;


public class Euler_3
{
    static long largeNumber = 600851475143L;

    public static void main (String[] args)
    {
        long start = System.nanoTime();
        long largestPrimeFactor = 0;

        for (int i = 1; i < sqrt(largeNumber) + 1; i += 2)
            if (largeNumber % i == 0 && isPrime(i) && 
                i > largestPrimeFactor)  
                largestPrimeFactor = i;

        long stop = System.nanoTime();
        System.out.println("Answer: " + largestPrimeFactor);
        System.out.printf("Time: %.4f\n", ((float) stop - start) / 1_000_000_000);
    }

    private static boolean isPrime(int n)
    {
        if (n < 2) return false;
        else if (n == 2) return true;

        int endRange = ((Double) ceil(sqrt(n))).intValue();
        IntStream range = IntStream.range(3, endRange);
        return range.allMatch(x -> n % x != 0);
    }
}
